napari/utils/tree/selection.py

```diff
@@ -0,0 +1,21 @@
+from typing import Iterable, Optional, TypeVar
+
+from ...utils.events import EventedSet
+
+_T = TypeVar("_T")
+
+
+class Selection(EventedSet[_T]):
+    def __init__(self, data: Iterable[_T] = (), current: Optional[_T] = None):
+        super().__init__(data=data)
+        self.events.add(current=None)
+        self._current = current
+
+    @property
```

Comment:
I wonder how we'll serialize this object too, it's not so easy as it is both a set and has an attribute. We've already hit that problem with the `Selection` set and so maybe we'll never even hit this object during serialization (and if we did, then we'd just ignore `current`), but just thought i'd ask

Reply:
> I wonder how we'll serialize this object too, 

After #2297, this is easy with
```python
    def _json_encode(self):
        """Return an object that can be used by json.dumps."""
        return {'data': list(self), 'current': self.current}
```

which is easily deserialized in the `validate` method:
```python
    def validate(cls, v, field: 'ModelField'):
        if isinstance(v, dict):
            data = v.get("data", [])
            current = v.get("current", None)
```

Reply:
very nice, i see how #2297 makes this much nicer